### DISCUSSION TITLE
Menambahkan kolom baru (dgn kriteria selisih waktu) di R Studio

### DISCUSSION THREAD

**Discussion by yohanespm77** [teamalgoritma/community#556] [Q&A] [answered]
Saya punya data seperti ini :
<img width="257" alt="image" src="https://user-images.githubusercontent.com/106010284/220249257-1e4321f0-6c10-4c81-bcd7-fe4cedfe0667.png">

Bagaimana cara membuat kolom baru yang isinya keterangan dari variabel "trx_time" dengan ketentuan :

- apabila di hari yg sama dan nama yg sama ada jam yg selisihnya kurang dari 30 detik maka didefine sebagai "1"
- apabila ada data berikutnya di hari yg sama dan nama yg sama selisih jamnya kurang dari 30 detik lagi, maka didefine sebagai "2", dst
- diluar itu didefine sebagai "0".

Outputnya mau seperti ini :
<img width="432" alt="image" src="https://user-images.githubusercontent.com/106010284/220250029-a28ec96e-012a-4c74-acee-7a4b08c1ee52.png">

Mohon bantuannya kak, terima kasih.

**Comment by finesaaa** [comment]
Dear pak @yohanespm77,
Terima kasih atas pertanyaannya, akan kami bantu carikan referensinya seperti biasa ya pak

**Reply by yohanespm77** [reply]
Baik kak.

**Comment by finesaaa** [comment, marked as answer]
Dear pak @yohanespm77,
Silahkan dapat diikuti dengan pembuatan fungsi kondisi dengan looping seperti kemarin, namun lumayan ada perbedaan, berikut
```
library(lubridate)
library(dplyr)
options(dplyr.summarise.inform = FALSE)

# create group df for counter
counter <- data %>% group_by(date=as_date(trx_time), name) %>% summarise()
counter['count'] <- rep(0, nrow(counter))
counter['first_hms'] <- rep(as.POSIXct('2024-01-01 00:00:00'), nrow(counter))
less_than_30s <- c()

# create looping for `less_than_30s` columns
for (row in 1:nrow(data)) {
  tmp <- data[row,]
  cond <- (tmp$name == counter$name & as_date(tmp$trx_time) == counter$date)
  print(tmp[cond, ]$trx_time)
  if (counter[cond, ]$first_hms == as.POSIXct('2024-01-01 00:00:00')) {
      counter[cond, ]$first_hms <- data[row,]$trx_time
      less_than_30s<- append(less_than_30s, 0)
  }
  else {
    if (difftime(hms::as.hms(tmp$trx_time), hms::as.hms(counter[cond, ]$first_hms )) <= 30) {
      counter[cond, ]$count <- counter[cond, ]$count + 1
      
      less_than_30s <- append(less_than_30s, counter[cond, ]$count)
    } else {
      less_than_30s<- append(less_than_30s, 0)
    }
  }
}

data['less_than_30s'] <- less_than_30s
data
```
![image](https://user-images.githubusercontent.com/57583780/220325705-8e5b0734-7311-4687-be5d-fb2f5b75d034.png)

**Comment by yohanespm77** [comment]
Terima kasih sblmnya kak @finesaaa 

Saya coba run codenya kak, tapi msh belum sesuai kak.
Ini ada jam yg selisihnya lebih dari 30 detik, tapi malah kedefine sebagai "1", harusnya "0". 

<img width="569" alt="image" src="https://user-images.githubusercontent.com/106010284/220345501-c3eaa633-be7d-4e54-b29d-fd32f5897e11.png">

<img width="451" alt="image" src="https://user-images.githubusercontent.com/106010284/220346794-b22f99b4-2f5c-401f-a7e8-621601af9d38.png">
Yg ini juga kak. Harusnya jam 03:11:03 itu kedefine sebagai "0",  jam 03:11:05 kedefine sebagai "1". Jika ada lagi di bawahnya, misalkan jam 03:11:20 maka akan kedefine sebagai "2", dst. 
Selisihnya dengan jam sebelumnya kak.

**Reply by finesaaa** [reply]
Baik pak Yohanes, terkait hal itu boleh Bapak coba otak-atik pada bagian 'if-else' nya 😄
Hasilnya boleh dilampirkan disini.

**Reply by yohanespm77** [reply]
Baik kak, terima kasih banyak kak @finesaaa atas bantuannya.
